napari/layers/utils/_slice_input.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+import warnings
+from dataclasses import dataclass
+from typing import List, Tuple, Union
+
+import numpy as np
+
+from napari.components.dims import reorder_after_dim_reduction
+from napari.utils.transforms import Affine
+from napari.utils.translations import trans
+
+
+@dataclass(frozen=True)
+class _SliceInput:
```

Comment:
I'm wondering about dataclass vs pydantic model here - would a model with `allow_mutation=False` in the config make more sense in terms of consistency with other parts of the codebase?

Reply:
My reason to use the built-in `dataclass` over pydantic's `BaseModel` here is that it's enough. `_SliceInput` is intended to be a private class that groups together a few attributes that we want to reason about together. We don't need validation (because the input is not coming from the user) or (de)serialization, so there doesn't seem to be a strong reason to jump to `BaseModel`.

Even though performance isn't a big motivator here, pydantic is about 20x slower than the built-in dataclass to create this object for 3D display.

```python
In [1]: from dataclasses import dataclass

In [2]: @dataclass
   ...: class _SliceInput:
   ...:     ndisplay: int
   ...:     point: tuple[float, ...]
   ...:     order: tuple[int, ...]
   ...: 

In [3]: %timeit _SliceInput(ndisplay=3, point=(0, 0, 0), order=(3, 0, 1))
399 ns ± 3.54 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: from pydantic import BaseModel

In [5]: class _SliceInput(BaseModel):
   ...:     ndisplay: int
   ...:     point: tuple[float, ...]
   ...:     order: tuple[int, ...]
   ...: 

In [6]: %timeit _SliceInput(ndisplay=3, point=(0, 0, 0), order=(3, 0, 1))
10.6 µs ± 116 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

10 microseconds isn't much, but that is on a fairly modern laptop and these objects may be created fairly frequently (e.g. imagine moving the slider or panning/zooming around the canvas), so it's best to keep this as light as possible.

I kind of understand the argument of consistency, so I would be willing to bend here, but will likely stick with `dataclass` unless forced to change it.

Reply:
For completeness, I did also try `attrs`, which is maybe a little faster than the built-in `dataclass`.

```python
In [1]: from attrs import define

In [2]: @define
   ...: class _SliceInput:
   ...:     ndisplay: int
   ...:     point: tuple[float, ...]
   ...:     order: tuple[int, ...]
   ...: 

In [3]: %timeit _SliceInput(ndisplay=3, point=(0, 0, 0), order=(3, 0, 1))
368 ns ± 32.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

But it's hard to justify it as a new explicit dependency based on that. It would also create more inconsistency in terms of when we should use `dataclasses`, `attrs`, and `pydantic`.

Reply:
Thank you! Totally agree on all points. 

Nice to see that the timing results align with the reported comparisons between `dataclass`, `attrs`, and `pydantic.BaseModel`. From what I see about `attrs` I might choose it instead of `pydantic` when making an initial dependency selection: get the speed of `dataclass` plus ability to turn on validations later down the road.

Reply:
> (e.g. imagine moving the slider or panning/zooming around the canvas),

This shouldn't happen on panning/zooming, only on dims changes. But yeah, I agree that dataclass is completely fine here.